### ISSUE TITLE
500 is caused when receiving a validation error on a child model

### ISSUE THREAD

**Issue by montague** [elixir-lang/ecto#1159]
My params map in a simple phoenix app looks like this:

```
%{
  "_csrf_token" => "B2c4PAN3DhkYVW4tG0wyQgddkdkegAA7RpxU1mXQ3Ya+zY1Oz2xNQ==",
  "_method" => "put",
  "_utf8" => "✓",
  "goal" => %{
    "description" => "build this app",
    "tags" => %{
        "0" => %{  # VALID
          "delete" => "false",
          "id" => "1",
          "tag_id" => "1",
          "goal_id" => "1"
        },
        "1" => %{ # INVALID (same tag_id & goal_id as above)
          "delete" => "false",
          "id" => "2",
          "tag_id" => "1",
          "goal_id" => "1"
        },
        "2" => %{ # VALID
          "delete" => "false",
          "id" => "8",
          "tag_id" => "3",
          "goal_id" => "1"
        }
      }
    },
  "id" => "1"
}
```

I have a uniqueness constraint across `tag_id` and `goal_id`. Here, I am submitting three tags to be updated on a `goal`. 
* `%{"0" => %{"delete" => "false", "id" => "1", "tag_id" => "1", "goal_id" => "1"}}` VALID
* `%{"1" => %{"delete" => "false", "id" => "2", "tag_id" => "1", "goal_id" => "1"}}` INVALID (breaks uniqueness constraint because `tag_id` and `goal_id` are the same as above.
* `"2" => %{"delete" => "false", "id" => "8", "tag_id" => "3", "goal_id" => "1"}}` VALID

I'd expect this submission to return a validation error from the uniqueness constraint. Instead, the following error is thrown, causing a 500:
```ERROR (in_failed_sql_transaction): current transaction is aborted, commands ignored until end of transaction block```

Here is the relevant section from server log in phoenix, where we try to update the tag associations with the goal:
```
[debug] BEGIN [] OK query=0.1ms
[debug] UPDATE "goals" SET "updated_at" = $1 WHERE "id" = $2 [{{2015, 12, 21}, {2, 52, 3, 0}}, 1] OK query=0.2ms
[debug] UPDATE "tagged_goals" SET "updated_at" = $1 WHERE "id" = $2 [{{2015, 12, 21}, {2, 52, 3, 0}}, 1] OK query=0.3ms
[debug] UPDATE "tagged_goals" SET "updated_at" = $1, "tag_id" = $2 WHERE "id" = $3 [{{2015, 12, 21}, {2, 52, 3, 0}}, 1, 2] ERROR query=0.3ms
[debug] UPDATE "tagged_goals" SET "updated_at" = $1, "tag_id" = $2 WHERE "id" = $3 [{{2015, 12, 21}, {2, 52, 3, 0}}, 3, 8] ERROR query=0.1ms
[debug] ROLLBACK [] OK query=0.1ms
[info] Sent 500 in 19ms
[error] #PID<0.640.0> running GSD.Endpoint terminated
```

I think this is a raw postgres error, pretty much the same as what's described in this SO article: http://stackoverflow.com/a/13103690

TLDR; I think we are getting an error inside a postgres transaction, then trying to continue with further operations inside that transaction. Ecto handles the first error, but not the subsequent one.

My expectation is that this situation would cause ecto to throw a typical validation error, not a 500. The exception itself is happening inside the `Repo.update(changeset)` statement in the controller.

I didn't see anything similar in the existing open issues, but apologies if I missed something.

**Comment by ericmj** [comment]
Have you added the constraint to the changeset? Can you show your changeset code and how you call it and the repo?

**Comment by montague** [comment]
Sure! Here's the changeset on the model:
```
  def changeset(model, params \\ :empty) do
    model
    |> cast(params, @required_fields, @optional_fields)
    |> unique_constraint(:goal_id, name: :tagged_goals_goal_id_tag_id_index,
                                    message: "is a duplicate")
    |> mark_for_deletion
  end
```

Here's the code that calls `update` in the controller:
```
def update(conn, %{"id" => id, "goal" => goal_params}) do
    goal = Repo.get!(Goal, id)
            |> Repo.preload(:sub_goals)
            |> Repo.preload(:tags)

    changeset = Goal.changeset(goal, goal_params)

    case Repo.update(changeset) do
      {:ok, goal} ->
        conn
        |> put_flash(:info, "Goal updated successfully.")
        |> redirect(to: goal_path(conn, :show, goal))
      {:error, changeset} ->
        render(conn, "edit.html", goal: goal, changeset: changeset)
    end
  end
```

Is that what you were looking for?

**Comment by montague** [comment]
awesome!!
